sky/utils/vpn_utils.py

```
@@ -0,0 +1,309 @@
+        return env_vars
+
+    def get_private_ip(self, hostname: str) -> Dict[str, str]:
+        """Get the private IP address from the hostname."""
+        private_ips = {}
+        for vpn_name, vpn_config in self._vpn_configs.items():
+            private_ip = vpn_config.get_private_ip(hostname)
+            private_ips[vpn_name] = private_ip
+        return private_ips
+
+    def get_private_ip_vpn(self, hostname: str, vpn_name: str) -> str:
+        """Get the private IP address from the hostname for a specific VPN."""
+        return self._vpn_configs[vpn_name].get_private_ip(hostname)
+
+    def remove_host(self, hostname: str) -> None:
+        """Remove a host from the VPN."""
+        for vpn_config in self._vpn_configs.values():
+            vpn_config.remove_host(hostname)
+
+    def to_yaml_config(self) -> Dict[str, Any]:
+        """Get the VPN configuration in YAML format."""
+        vpn_config_dict = {}
+        for vpn_name, vpn_config in self._vpn_configs.items():
+            vpn_yaml = vpn_config.to_yaml_config()
+            if vpn_yaml:
+                vpn_config_dict[vpn_name] = vpn_yaml
+            else:
+                vpn_config_dict[vpn_name] = True
+        return vpn_config_dict
+
+    def to_backend_config(self) -> Dict[str, Any]:
+        """Generate the VPN configuration for the backend."""
+        vpn_config_dict = {}
+        for vpn_name, vpn_config in self._vpn_configs.items():
+            vpn_config_dict[vpn_name] = vpn_config.to_backend_config()
+        return vpn_config_dict
+
+    def __getitem__(self, key: str) -> VPN:
+        return self._vpn_configs[key]
+
+    def __setitem__(self, key: str, value: VPN) -> None:
+        self._vpn_configs[key] = value
+
+
+def rewrite_cluster_info_by_vpn(
+    cluster_info: common.ClusterInfo,
+    vpn_config: VPNConfig,
+) -> common.ClusterInfo:
+    for (instance_id, instance_list) in cluster_info.instances.items():
+        # TODO(yi): test if this works on TPU VM.
```

Lets raise an error for TPU VM + vpn config first. Also pls submit an issue for that ;)